### PR TITLE
Support 2D inner reduction scheduler with autotuning

### PR DESCRIPTION
**Pull request by rdspring1**
# Summary:
This PR creates an SOL autotuning script for the 2d inner reduction scheduler. It trains a random forest to predict the best performing configuration for the reduction scheduler.

# Inner Reduction Fusions:
1. Sum --- `y = sum(x, dim=-1)`
2. Add Sum --- `z = sum(x1 + x2 + x3 + x4, dim=-1)`
3. Tanh Sum --- `y = sum(tanh(x), dim=-1)`
4. Exp Sum --- `z = sum(exp(x), dim=-1)`

### CODE REVIEW

**Comment by rdspring1** [comment]
!test

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

File: doc/dev/python_scheduling/autotune_inner_reduction.py [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+
+class AutotuneInnerReduction:
+    class FUSION(Enum):
+        INNER_SUM = 1
+        ADD_SUM = 2
+        TANH_SUM = 3
+        EXP_SUM = 4
+
+    @dataclass(unsafe_hash=True)
+    class InnerReductionConfiguration:
```
Comment (liqiangxl) [inline doc/dev/python_scheduling/autotune_inner_reduction.py:72]: add comment about how the generated kernel changes when different paras are used.

File: doc/dev/python_scheduling/autotune_inner_reduction.py [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+        grid_x_limit = 2147483647
+        grid_y_limit = 65535
+
+        reduction_params.schedule_3D = False
+        reduction_params.fastest_dim = True
+        reduction_params.cross_block_inner_reduction = True
+        reduction_params.block_dim_inner_reduction = ParallelType.block_x
+        reduction_params.cross_grid_inner_reduction = scheduler_config.grdim > 1
+        reduction_params.multiple_reds_per_blk = scheduler_config.bdimy > 1
+        reduction_params.pad_inner_reduction_to_warp = (
```
Comment (liqiangxl) [inline doc/dev/python_scheduling/autotune_inner_reduction.py:102]: should be `bdimx >= warp && bdimx * bdimy <= max threads per block && bdimx % warp ==0`

File: doc/dev/python_scheduling/autotune_inner_reduction.py [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+            scheduler_config.bdimx > warp_size
+        ) and (
+            (scheduler_config.bdimx * scheduler_config.bdimy)
+            < max_number_of_threads_cta
+        )
+        reduction_params.unroll_factor_inner_reduction = (
```
Comment (liqiangxl) [inline doc/dev/python_scheduling/autotune_inner_reduction.py:108]: Looks like `unroll_on_top_of_vectorization` is missing

File: doc/dev/python_scheduling/autotune_inner_reduction.py [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+
+        reduction_params.lparams.bdimx = scheduler_config.bdimx
+        reduction_params.lparams.bdimy = scheduler_config.bdimy
+
+    # For reduction scheduler, we test the cartesian product of vectorization and
+    # unroll factors.
+    def generate_scheduler_configurations(self, input_shape):
+        threads_per_cta_options = [128, 256, 512, 1024]
+        vectorization_factor_options = [1, 2, 4, 8]
+        unroll_factor_options = list(range(1, 11))
```
Comment (liqiangxl) [inline doc/dev/python_scheduling/autotune_inner_reduction.py:154]: we have `iteration unroll` and `reduction unroll on top of vectorization`

**Review by liqiangxl — COMMENTED** [review]
LGTM. Similar to pointwise, recomment to use variables instead of instant numbers which provides a more clear interface to the user.

File: doc/dev/python_scheduling/autotune_inner_reduction.py (hunk 3 above)
Comment (rdspring1) [reply doc/dev/python_scheduling/autotune_inner_reduction.py:108]: It wasn't available when I wrote the github script. I implemented in a separate branch.

Comment (rdspring1) [reply doc/dev/python_scheduling/autotune_inner_reduction.py:108]: I'll add it in a separate PR.

File: doc/dev/python_scheduling/autotune_inner_reduction.py (hunk 4 above)
Comment (rdspring1) [reply doc/dev/python_scheduling/autotune_inner_reduction.py:154]: I'll add it in a separate PR.

File: doc/dev/python_scheduling/autotune_inner_reduction.py (hunk 2 above)
Comment (rdspring1) [reply doc/dev/python_scheduling/autotune_inner_reduction.py:102]: In the cpp code, I see:
```cpp
bool pad_bdimx = bdimx > 16 && bdimx * bdimy < threads_per_block;
```
and
```cpp
  bool pad_bdimx = bdimx > 16 &&
      bdimx * bdimy <
          (int64_t)at::cuda::getCurrentDeviceProperties()->maxThreadsPerBlock;
```

File: doc/dev/python_scheduling/autotune_inner_reduction.py (hunk 4 above)
Comment (rdspring1) [reply doc/dev/python_scheduling/autotune_inner_reduction.py:164]: https://github.com/NVIDIA/Fuser/pull/3487

File: doc/dev/python_scheduling/autotune_inner_reduction.py (hunk 3 above)
Comment (rdspring1) [reply doc/dev/python_scheduling/autotune_inner_reduction.py:118]: https://github.com/NVIDIA/Fuser/pull/3487

**Comment by rdspring1** [comment]
!test

**Comment by rdspring1** [comment]
!build

**Review by liqiangxl — APPROVED** [review]
LGTM.

File: doc/dev/python_scheduling/autotune_inner_reduction.py [hunk 5]
```diff
@@ -0,0 +1,405 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024-present NVIDIA CORPORATION & AFFILIATES.
+# All rights reserved.
+# SPDX-License-Identifier: BSD-3-Clause
+# Owner(s): ["module: nvfuser"]
```
Comment (jacobhinkle) [inline doc/dev/python_scheduling/autotune_inner_reduction.py:4]: I'm curious: I noticed this `# Owner(s): ["module: nvfuser"]` line in other python files in our repo. Is this a standard and is it used by some tool?

**Review by jacobhinkle — APPROVED** [review]

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Pull request merged**